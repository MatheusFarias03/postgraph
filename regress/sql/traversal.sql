/*
 * Copyright (C) 2023 PostGraphDB
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 */

LOAD 'postgraph';
SET search_path TO postgraph;

SELECT create_graph('variable_edge');
SELECT create_vlabel('variable_edge', 'vlabel');
SELECT create_vlabel('variable_edge', 'elabel');

--Basic Route Creation
SELECT build_traversal(
	build_vertex(_graphid(3, 2), graphid, gtype_build_map()),
	build_edge(_graphid(4, 1), '2'::graphid, '3'::graphid,  graphid, gtype_build_map()),
	build_vertex(_graphid(3, 3),  graphid, gtype_build_map())
)
FROM ag_graph;

SELECT build_traversal(
        build_vertex(_graphid(3, 2), graphid, gtype_build_map()),
        build_edge(_graphid(4, 1), '2'::graphid, '3'::graphid,  graphid, gtype_build_map()),
        build_vertex(_graphid(3, 3), graphid, gtype_build_map()),
	build_edge(_graphid(4, 4), '3'::graphid, '5'::graphid,  graphid, gtype_build_map()),
        build_vertex(_graphid(3, 5),  graphid, gtype_build_map())
)
FROM ag_graph;

SELECT build_traversal(
        build_vertex(_graphid(3, 2), graphid, gtype_build_map()),
        build_edge(_graphid(4, 1), '2'::graphid, '3'::graphid,  graphid, gtype_build_map()),
        build_vertex(_graphid(3, 3), graphid, gtype_build_map()),
        build_vertex(_graphid(3, 5),  graphid, gtype_build_map())
)
FROM ag_graph;

SELECT build_traversal(
        build_vertex(_graphid(3, 2), graphid, gtype_build_map()),
        build_edge(_graphid(4, 1), '2'::graphid, '3'::graphid,  graphid, gtype_build_map()),
        build_edge(_graphid(4, 4), '3'::graphid, '5'::graphid,  graphid, gtype_build_map()),
        build_vertex(_graphid(3, 5),  graphid, gtype_build_map())
)
FROM ag_graph;

SELECT build_traversal(
        build_vertex(_graphid(3, 2), graphid, gtype_build_map()),
        build_edge(_graphid(4, 1), '2'::graphid, '3'::graphid,  graphid, gtype_build_map()),
        build_vertex(_graphid(3, 3), graphid, gtype_build_map()),
        NULL,
        build_vertex(_graphid(3, 5),  graphid, gtype_build_map())
)
FROM ag_graph;


SELECT build_traversal(
        build_vertex(_graphid(3, 2), graphid, gtype_build_map()),
        build_variable_edge(
		build_edge(_graphid(4, 1), '2'::graphid, '3'::graphid,  graphid, gtype_build_map())
	),
        build_vertex(_graphid(3, 3),  graphid, gtype_build_map())
)
FROM ag_graph;

SELECT build_traversal(
        build_vertex(_graphid(3, 2), graphid, gtype_build_map()),
        build_variable_edge(
            build_edge(_graphid(4, 1), '2'::graphid, '3'::graphid,  graphid, gtype_build_map()),
            build_vertex(_graphid(3, 3), graphid, gtype_build_map()),
            build_edge(_graphid(4, 4), '3'::graphid, '5'::graphid,  graphid, gtype_build_map())
        ),
        build_vertex(_graphid(3, 3),  graphid, gtype_build_map())
)
FROM ag_graph;

SELECT drop_graph('variable_edge', true);
