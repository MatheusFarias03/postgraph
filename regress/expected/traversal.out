/*
 * Copyright (C) 2023 PostGraphDB
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 */
LOAD 'postgraph';
SET search_path TO postgraph;
SELECT create_graph('variable_edge');
NOTICE:  graph "variable_edge" has been created
 create_graph 
--------------
 
(1 row)

SELECT create_vlabel('variable_edge', 'vlabel');
NOTICE:  VLabel "vlabel" has been created
 create_vlabel 
---------------
 
(1 row)

--Basic Route Creation
SELECT build_traversal(
	build_vertex(_graphid(3, 2), graphid, gtype_build_map()),
	build_edge('1'::graphid, '2'::graphid, '3'::graphid, $$edge_label$$, gtype_build_map()),
	build_vertex(_graphid(3, 3),  graphid, gtype_build_map())
)
FROM ag_graph;
                                                                                               build_traversal                                                                                                
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 [{"id": 844424930131970, "label": "vlabel", "properties": {}}, {"id": 1, "start_id": 2, "end_id": 3, "label": "edge_label", "properties": {}}, {"id": 844424930131971, "label": "vlabel", "properties": {}}]
(1 row)

SELECT build_traversal(
        build_vertex(_graphid(3, 2), graphid, gtype_build_map()),
        build_edge('1'::graphid, '2'::graphid, '3'::graphid, $$edge_label$$, gtype_build_map()),
        build_vertex(_graphid(3, 3), graphid, gtype_build_map()),
	build_edge('4'::graphid, '3'::graphid, '5'::graphid, $$edge_label$$, gtype_build_map()),
        build_vertex(_graphid(3, 5),  graphid, gtype_build_map())
)
FROM ag_graph;
                                                                                                                                                                      build_traversal                                                                                                                                                                       
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 [{"id": 844424930131970, "label": "vlabel", "properties": {}}, {"id": 1, "start_id": 2, "end_id": 3, "label": "edge_label", "properties": {}}, {"id": 844424930131971, "label": "vlabel", "properties": {}}, {"id": 4, "start_id": 3, "end_id": 5, "label": "edge_label", "properties": {}}, {"id": 844424930131973, "label": "vlabel", "properties": {}}]
(1 row)

SELECT build_traversal(
        build_vertex(_graphid(3, 2), graphid, gtype_build_map()),
        build_edge('1'::graphid, '2'::graphid, '3'::graphid, $$edge_label$$, gtype_build_map()),
        build_vertex(_graphid(3, 3), graphid, gtype_build_map()),
        build_vertex(_graphid(3, 5),  graphid, gtype_build_map())
)
FROM ag_graph;
ERROR:  arguement 3 build_traversal() must be an edge
SELECT build_traversal(
        build_vertex(_graphid(3, 2), graphid, gtype_build_map()),
        build_edge('1'::graphid, '2'::graphid, '3'::graphid, $$edge_label$$, gtype_build_map()),
        build_edge('4'::graphid, '3'::graphid, '5'::graphid, $$edge_label$$, gtype_build_map()),
        build_vertex(_graphid(3, 5),  graphid, gtype_build_map())
)
FROM ag_graph;
ERROR:  arguement 2 build_traversal() must be a vertex
SELECT build_traversal(
        build_vertex(_graphid(3, 2), graphid, gtype_build_map()),
        build_edge('1'::graphid, '2'::graphid, '3'::graphid, $$edge_label$$, gtype_build_map()),
        build_vertex(_graphid(3, 3), graphid, gtype_build_map()),
        NULL,
        build_vertex(_graphid(3, 5),  graphid, gtype_build_map())
)
FROM ag_graph;
 build_traversal 
-----------------
 
(1 row)

SELECT build_traversal(
        build_vertex(_graphid(3, 2), graphid, gtype_build_map()),
        build_variable_edge(
		build_edge('1'::graphid, '2'::graphid, '3'::graphid, $$edge_label$$, gtype_build_map())
	),
        build_vertex(_graphid(3, 3),  graphid, gtype_build_map())
)
FROM ag_graph;
                                                                                               build_traversal                                                                                                
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 [{"id": 844424930131970, "label": "vlabel", "properties": {}}, {"id": 1, "start_id": 2, "end_id": 3, "label": "edge_label", "properties": {}}, {"id": 844424930131971, "label": "vlabel", "properties": {}}]
(1 row)

SELECT build_traversal(
        build_vertex(_graphid(3, 2), graphid, gtype_build_map()),
        build_variable_edge(
            build_edge('1'::graphid, '2'::graphid, '3'::graphid, $$edge_label$$, gtype_build_map()),
            build_vertex(_graphid(3, 3), graphid, gtype_build_map()),
            build_edge('4'::graphid, '3'::graphid, '5'::graphid, $$edge_label$$, gtype_build_map())
        ),
        build_vertex(_graphid(3, 3),  graphid, gtype_build_map())
)
FROM ag_graph;
                                                                                                                                                                      build_traversal                                                                                                                                                                       
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 [{"id": 844424930131970, "label": "vlabel", "properties": {}}, {"id": 1, "start_id": 2, "end_id": 3, "label": "edge_label", "properties": {}}, {"id": 844424930131971, "label": "vlabel", "properties": {}}, {"id": 4, "start_id": 3, "end_id": 5, "label": "edge_label", "properties": {}}, {"id": 844424930131971, "label": "vlabel", "properties": {}}]
(1 row)

SELECT drop_graph('variable_edge', true);
NOTICE:  drop cascades to 3 other objects
DETAIL:  drop cascades to table variable_edge._ag_label_vertex
drop cascades to table variable_edge._ag_label_edge
drop cascades to table variable_edge.vlabel
NOTICE:  graph "variable_edge" has been dropped
 drop_graph 
------------
 
(1 row)

